---
title: "blast_annotation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Functions

```{r functions}
### preprocess the repeatMaster output
preprocessing_repeats <- function(repeat_output)
{
  # prepare right positions
  repeat_output$query_left[repeat_output$query_left < 0] <- paste("(", 
                                                                  repeat_output$query_left[repeat_output$query_left < 0]*-1,
                                                                  ")", sep = "")
  repeat_output$query_left[repeat_output$query_left == 0] <-  paste("(", 
                                                                  repeat_output$query_left[repeat_output$query_left == 0] ,
                                                                  ")", sep = "")
  repeat_output$position_begin[repeat_output$position_begin < 0] <- paste("(", 
                                                                  repeat_output$position_begin[repeat_output$position_begin < 0]*-1,
                                                                  ")", sep = "")
  repeat_output$position_begin[repeat_output$position_begin == 0] <-  paste("(", 
                                                                  repeat_output$position_begin[repeat_output$position_begin == 0] ,
                                                                  ")", sep = "")
  repeat_output$repeat_left[repeat_output$repeat_left < 0] <- paste("(", 
                                                                  repeat_output$repeat_left[repeat_output$repeat_left < 0]*-1,
                                                                  ")", sep = "")
  repeat_output$repeat_left[repeat_output$repeat_left == 0] <-  paste("(", 
                                                                  repeat_output$repeat_left[repeat_output$repeat_left == 0] ,
                                                                  ")", sep = "")
  ## define right and left positions

  repeat_colnames_ordered <- c("SW_score", "perc_div","perc","perc_ins","insertions_ID",
                              "insertion_position_begin", "insertion_position_end", "insertion_position_left",
                              "strand", "matching_transposon", "transposon_class_family", 
                              "transposon_position_left","transposon_position_begin", "transposon_position_end", "transposon_position_right")

  
  colnames(repeat_output) <- c("SW_score", "perc_div","perc","perc_ins","insertions_ID",
                              "insertion_position_begin", "insertion_position_end", "insertion_position_left",
                              "strand", "matching_transposon", "transposon_class_family", 
                              "transposon_position_begin", "transposon_position_end", "transposon_position_left",
                              "alignment_ID")
 
  trast_left <- repeat_output$transposon_position_left
  repeat_output$transposon_position_right <- NA
  repeat_output$transposon_position_left <- NA
  repeat_output <- repeat_output[,repeat_colnames_ordered] 
  
  ind <- which(repeat_output$strand == "C")

  repeat_output$transposon_position_left[ind] <- repeat_output$transposon_position_begin[ind] 
  repeat_output$transposon_position_left[-ind] <- "(0)"
  
  repeat_output$transposon_position_begin[ind] <- repeat_output$transposon_position_end[ind]
  repeat_output$transposon_position_end[ind] <- trast_left[ind]
  
  repeat_output$transposon_position_right[ind] <- "(0)"
  repeat_output$transposon_position_right[-ind] <- trast_left[-ind]
  
  return(repeat_output)
}

#define insertion position and covarage

define_insertion_coverage <- function(repeat_output, vcf)
{
  # add insertion information
  repeat_output$transposon_insertions_cover <- "partly_insertion"
  repeat_output$transposon_insertions_cover[repeat_output$strand == "+" & 
                                       repeat_output$transposon_position_begin == "1" &
                                       repeat_output$transposon_position_right == "(0)"] <- "full_insertion"
  
  repeat_output$transposon_insertions_cover[repeat_output$strand == "C" & 
                                       repeat_output$transposon_position_left == "(0)" &
                                       repeat_output$transposon_position_end == "1"] <- "full_insertion"
  # add covarage

  repeat_output_cov <- as.data.frame(vcf@fix[which(vcf@fix[,"ID"] %in% repeat_output$insertions_ID),c("ID","INFO" )])
  
  repeat_output_cov$insertions_coverage <- sapply(repeat_output_cov$INFO, function(x)
  {
  x <- strsplit(strsplit(x, ";")[[1]][2], "=")[[1]][2]
  return(x)
  })
  
  repeat_output_cov <- repeat_output_cov[,c("ID", "insertions_coverage")]
  colnames(repeat_output_cov)[1] <- "insertions_ID"  
  repeat_output <- merge(repeat_output, repeat_output_cov, by = "insertions_ID")
  repeat_output <- repeat_output[order(as.numeric(repeat_output$insertions_coverage), decreasing = T),]
  
  return(repeat_output)
  
}


## define gene
define_gene <- function(insertion, genes)
{
  insertion$insertion_position <- as.numeric( insertion$insertion_position)
  insertion_gene <- data.frame(insertions_ID = unique(insertion$insertions_ID ))
  insertion_gene_full_temp <- data.frame(insertions_ID = vector("character", 0),
                                         chr_gene = vector("character", 0),
                                         start = vector("numeric", 0),
                                         end = vector("numeric", 0),
                                         strand = vector("character", 0),
                                         gene_id = vector("character", 0), 
                                         gene_name = vector("character", 0))
  for (i in 1:nrow(insertion)) 
    {
      insertion_gene_temp <- genes[which(genes$chr == insertion$chr[i] & 
                                  genes$start <= insertion$insertion_position[i] &
                                  genes$end >= insertion$insertion_position[i]),c("chr", "start", "end", "strand", "gene_id", "gene_name")]
      if(nrow(insertion_gene_temp) > 0)
      {
      insertion_gene_temp$insertions_ID <- insertion$insertions_ID[i]
      insertion_gene_full_temp <- rbind(insertion_gene_full_temp,insertion_gene_temp )
      
      }
  }

  insertion_gene <- merge(insertion_gene, insertion_gene_full_temp, by = "insertions_ID")
  return(insertion_gene)

}

```



## Annotate blast results

Load files

```{r load_files}
library("readxl")
library("vcfR")

mm1s_intersect <- read_xlsx("../result/insertion_repeatmasker/MM1S_insertions.xlsx", sheet = "MM1S_intersect")
mm1s_match_intersect <- read_xlsx("../result/insertion_repeatmasker/MM1S_insertions.xlsx", sheet = "MM1S_match_filtered")
mm1s_mismatch_intersect <- read_xlsx("../result/insertion_repeatmasker/MM1S_insertions.xlsx", sheet = "MM1S_mismatch_filtered")

mia_intersect <- read_xlsx("../result/insertion_repeatmasker/MIA_insertions.xlsx", sheet = "MIA_intersect")
mia_match_intersect  <- read_xlsx("../result/insertion_repeatmasker/MIA_insertions.xlsx", sheet = "MIA_match_filtered")
mia_mismatch_intersect <- read_xlsx("../result/insertion_repeatmasker/MIA_insertions.xlsx", sheet = "MIA_mismatch_filtered")



mm1s_match_vcf <- read.vcfR("../data/pamir_vcf/MM1S_matchet_with_ref.vcf.gz")
mm1s_mismatch_vcf <- read.vcfR("../data/pamir_vcf/MM1S_mismatch_with_ref.vcf.gz")

mia_match_vcf <- read.vcfR("../data/pamir_vcf/MIA_matched_with_ref.vcf.gz")
mia_mismatch_vcf <- read.vcfR("../data/pamir_vcf/MIA_mismatch_with_ref.vcf.gz")

```


Process the data

```{r process_data}
#mm1s_intersect
mm1s_intersect <- preprocessing_repeats(mm1s_intersect)
mm1s_match_intersect <- preprocessing_repeats(mm1s_match_intersect)
mm1s_mismatch_intersect <- preprocessing_repeats(mm1s_mismatch_intersect)


mia_intersect <- preprocessing_repeats(mia_intersect)
mia_match_intersect <- preprocessing_repeats(mia_match_intersect)
mia_mismatch_intersect <- preprocessing_repeats(mia_mismatch_intersect)

```


#define overlap in match and mismatch

```{r define_overlap}

#mm1s
mm1s_match_ids <- sapply(mm1s_match_intersect$insertions_ID, function(x)
  {
  x <- paste(strsplit(x , "-")[[1]][1], strsplit(x, "-")[[1]][2], sep = "-")
  return(x)
  })
mm1s_mismatch_ids <-  sapply(mm1s_mismatch_intersect$insertions_ID, function(x)
  {
  x <- paste(strsplit(x , "-")[[1]][1], strsplit(x, "-")[[1]][2], sep = "-")
  return(x)
  })

ind <- intersect(mm1s_match_ids,mm1s_mismatch_ids )
 
mm1s_match_intersect <- mm1s_match_intersect[-which(mm1s_match_intersect$insertions_ID %in% 
                                                      names(mm1s_match_ids[which(mm1s_match_ids %in% mm1s_mismatch_ids)])),]
mm1s_mismatch_intersect <- mm1s_mismatch_intersect[-which(mm1s_mismatch_intersect$insertions_ID %in% 
                                                             names(mm1s_mismatch_ids[which(mm1s_mismatch_ids %in% mm1s_match_ids)])),]

#mia
mia_match_ids <- sapply(mia_match_intersect$insertions_ID, function(x)
  {
  x <- paste(strsplit(x , "-")[[1]][1], strsplit(x, "-")[[1]][2], sep = "-")
  return(x)
  })
mia_mismatch_ids <- sapply(mia_mismatch_intersect$insertions_ID, function(x)
  {
  x <- paste(strsplit(x , "-")[[1]][1], strsplit(x, "-")[[1]][2], sep = "-")
  return(x)
  })
ind <- intersect(mia_match_ids,mia_mismatch_ids )

ind <- names(mia_match_ids[which(mia_match_ids %in% mia_mismatch_ids)])
mia_match_intersect <- mia_match_intersect[-which(mia_match_intersect$insertions_ID %in% 
                                                    ind),]

ind <- names(mia_mismatch_ids[which(mia_mismatch_ids %in% mia_match_ids)])
mia_mismatch_intersect <- mia_mismatch_intersect[-which(mia_mismatch_intersect$insertions_ID %in% 
                                                    ind),]
```



Define full inserted positinos
Add covarage data

```{r add_covarage}

mm1s_intersect <- define_insertion_coverage(mm1s_intersect, mm1s_match_vcf)
mm1s_match_intersect <- define_insertion_coverage(mm1s_match_intersect, mm1s_match_vcf)
mm1s_mismatch_intersect <- define_insertion_coverage(mm1s_mismatch_intersect, mm1s_mismatch_vcf)

mia_intersect <- define_insertion_coverage(mia_intersect, mia_match_vcf)
mia_match_intersect <- define_insertion_coverage(mia_match_intersect, mia_match_vcf)
mia_mismatch_intersect <- define_insertion_coverage(mia_mismatch_intersect, mia_mismatch_vcf)


dir.create("../result/insertion_repeatmasker_annotation")

# From mm1s match and mismatch cell line is removed. mm1s_intersect - is intersect of match and mismatch files without cell line
write.table(mm1s_intersect, "../result/insertion_repeatmasker_annotation/mm1s_match_mismatch_interserct_no_cell_line_insertions_transposon.txt", sep = "\t", quote = F, row.names = F)
# From mm1s match cell line and mismatch are removed
write.table(mm1s_match_intersect, "../result/insertion_repeatmasker_annotation/mm1s_match_unique_insertions_no_cell_line_mismatch_transposon.txt", sep = "\t", quote = F, row.names = F)
# From mm1s mismatch cell line and match are removed
write.table(mm1s_mismatch_intersect, "../result/insertion_repeatmasker_annotation/mm1s_mismatch_unique_insertions_no_cell_line_match_transposon.txt", sep = "\t", quote = F, row.names = F)

# From mia match and mismatch cell line is removed. mm1s_intersect - is intersect of match and mismatch files without cell line
write.table(mia_intersect, "../result/insertion_repeatmasker_annotation/mia_match_mismatch_interserct_no_cell_line_insertions_transposon.txt", sep = "\t", quote = F, row.names = F)
# From mia match cell line and mismatch are removed
write.table(mia_match_intersect, "../result/insertion_repeatmasker_annotation/mia_match_unique_insertions_no_cell_line_mismatch_transposon.txt", sep = "\t", quote = F, row.names = F)
# From mia mismatch cell line and match are removed
write.table(mia_mismatch_intersect, "../result/insertion_repeatmasker_annotation/mia_mismatch_unique_insertions_no_cell_line_match_transposon.txt", sep = "\t", quote = F, row.names = F)

```

Calculate statistics 

```{r calc_stat}
### MM1S
#mm1s_intersect
mm1s_intersect_trans <- as.data.frame(table(mm1s_intersect$transposon_class_family)[order(table(mm1s_intersect$transposon_class_family), decreasing = T)]/nrow(mm1s_intersect))
colnames(mm1s_intersect_trans) <- c("transposon_family", "mm1s_match_mismatch_overlap_insertion_frequency")

#mm1s_match_intersect
mm1s_match_intersect_trans <- as.data.frame(table(mm1s_match_intersect$transposon_class_family)[order(table(mm1s_match_intersect$transposon_class_family), decreasing = T)]/nrow(mm1s_match_intersect))
colnames(mm1s_match_intersect_trans) <- c("transposon_family", "mm1s_match_unique_insertion_frequency")

#mm1s_mismatch_intersect
mm1s_mismatch_intersect_trans <- as.data.frame(table(mm1s_mismatch_intersect$transposon_class_family)[order(table(mm1s_mismatch_intersect$transposon_class_family), decreasing = T)]/nrow(mm1s_mismatch_intersect))
colnames(mm1s_mismatch_intersect_trans) <- c("transposon_family", "mm1s_mismatch_unique_insertion_frequency")

mm1s_transposones <- merge(mm1s_intersect_trans,mm1s_match_intersect_trans, by =  "transposon_family", all = T)
mm1s_transposones <- merge(mm1s_transposones,mm1s_mismatch_intersect_trans, by =  "transposon_family", all = T)

mm1s_transposones <- apply(mm1s_transposones, 2, function(x)
  {
  x[is.na(x)] <- 0
  return(x)
  })
### MIA
#mia_match_intersect 
mia_match_intersect_trans <- as.data.frame(table(mia_match_intersect$transposon_class_family)[order(table(mia_match_intersect$transposon_class_family), decreasing = T)]/nrow(mia_match_intersect))
colnames(mia_match_intersect_trans) <- c("transposon_family", "mia_match_unique_insertion_frequency")

#mia_mismatch_intersect
mia_mismatch_intersect_trans <-  as.data.frame(table(mia_mismatch_intersect$transposon_class_family)[order(table(mia_mismatch_intersect$transposon_class_family), decreasing = T)]/nrow(mia_mismatch_intersect))
colnames(mia_mismatch_intersect_trans) <- c("transposon_family", "mia_mismatch_unique_insertion_frequency")

#mia_intersect
mia_intersect_trans <-  as.data.frame(table(mia_intersect$transposon_class_family)[order(table(mia_intersect$transposon_class_family), decreasing = T)]/nrow(mia_intersect))
colnames(mia_intersect_trans) <- c("transposon_family", "mia_match_mismatch_overlap_insertion_frequency")

mia_transposones <- merge(mia_intersect_trans,mia_match_intersect_trans, by =  "transposon_family", all = T)
mia_transposones <- merge(mia_transposones,mia_mismatch_intersect_trans, by =  "transposon_family", all = T)

mia_transposones <- apply(mia_transposones, 2, function(x)
  {
  x[is.na(x)] <- 0
  return(x)
  })

### save

write.table(mm1s_transposones, "../result/insertion_repeatmasker_annotation/mm1s_transposon_frequency.txt", sep = "\t", row.names = F, quote = F)
write.table(mia_transposones, "../result/insertion_repeatmasker_annotation/mia_transposon_frequency.txt", sep = "\t", row.names = F, quote = F)
```

################ Gene annotation ################

Annotate genes in inserted positions


```{r gene_annotation}
gene_annotation <- read.table("../data/annotaion/Homo_sapiens.GRCh38.106_genes.gtf", sep = "\t")
colnames(gene_annotation) <- c("chr", "source","feature", "start", "end", "score", "strand", "frame", "attribute")

gene_annotation$gene_id <- sapply(gene_annotation$attribute, function(x)
  {
  x <- strsplit(x, ";")[[1]][grep("gene_id", strsplit(x, ";")[[1]])]
  x <- strsplit(x, " ")[[1]][2]
  return(x)
  })
gene_annotation$gene_name  <- sapply(gene_annotation$attribute, function(x)
  {
   x <- strsplit(x, ";")[[1]][grep("gene_name", strsplit(x, ";")[[1]])]
   if(length(x) > 0)
   {
    x <- strsplit(x, " ")[[1]][3]
    return(x)
   }
  return("")
  })
```

prepare insertion table

```{r prep_insertions}

### define insertions gene regions
#mm1s_intersect
mm1s_intersect_gene <- data.frame(insertions_ID = mm1s_intersect$insertions_ID,
                                  chr = sapply( mm1s_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][1]
                                    x <- gsub("chr", "", x)
                                    return(x)
                                    }),
                                  insertion_position = sapply( mm1s_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][2]
                                    return(x)
                                    })
                                  )

#mm1s_match_intersect
mm1s_match_intersect_gene <- data.frame(insertions_ID = mm1s_match_intersect$insertions_ID,
                                  chr = sapply( mm1s_match_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][1]
                                    x <- gsub("chr", "", x)
                                    return(x)
                                    }),
                                  insertion_position = sapply( mm1s_match_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][2]
                                    return(x)
                                    })
                                  )
#mm1s_mismatch_intersect
mm1s_mismatch_intersect_gene <- data.frame(insertions_ID = mm1s_mismatch_intersect$insertions_ID,
                                  chr = sapply( mm1s_mismatch_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][1]
                                    x <- gsub("chr", "", x)
                                    return(x)
                                    }),
                                  insertion_position = sapply( mm1s_mismatch_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][2]
                                    return(x)
                                    })
                                  )

#mia_intersect
mia_intersect_gene <- data.frame(insertions_ID = mia_intersect$insertions_ID,
                                  chr = sapply( mia_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][1]
                                    x <- gsub("chr", "", x)
                                    return(x)
                                    }),
                                  insertion_position = sapply( mia_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][2]
                                    return(x)
                                    })
                                  )
#mia_match_intersect
mia_match_intersect_gene <- data.frame(insertions_ID = mia_match_intersect$insertions_ID,
                                  chr = sapply( mia_match_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][1]
                                    x <- gsub("chr", "", x)
                                    return(x)
                                    }),
                                  insertion_position = sapply( mia_match_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][2]
                                    return(x)
                                    })
                                  )
#mia_mismatch_intersect
mia_mismatch_intersect_gene <- data.frame(insertions_ID = mia_mismatch_intersect$insertions_ID,
                                  chr = sapply( mia_mismatch_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][1]
                                    x <- gsub("chr", "", x)
                                    return(x)
                                    }),
                                  insertion_position = sapply( mia_mismatch_intersect$insertions_ID, function(x)
                                    {
                                    x <- strsplit(x, "-")[[1]][2]
                                    return(x)
                                    })
                                  )

```



Define gene

```{r define_gene}

### define genies in insertions 
#mm1s_intersect
mm1s_intersect_gene_annotations <- define_gene(insertion = mm1s_intersect_gene,
                                               genes = gene_annotation) 
mm1s_intersect_gene_annotations <- mm1s_intersect_gene_annotations[!duplicated(mm1s_intersect_gene_annotations),]

#mm1s_match_intersect
mm1s_match_intersect_gene_annotaions <- define_gene(insertion = mm1s_match_intersect_gene,
                                               genes = gene_annotation) 
mm1s_match_intersect_gene_annotaions <- mm1s_match_intersect_gene_annotaions[!duplicated(mm1s_match_intersect_gene_annotaions),]

#mm1s_mismatch_intersect
mm1s_mismatch_intersect_gene_annotations<- define_gene(insertion = mm1s_mismatch_intersect_gene,
                                               genes = gene_annotation) 
mm1s_mismatch_intersect_gene_annotations <- mm1s_mismatch_intersect_gene_annotations[!duplicated(mm1s_mismatch_intersect_gene_annotations),]

#mia_intersect
mia_intersect_gene_annotations<- define_gene(insertion = mia_intersect_gene,
                                               genes = gene_annotation) 
mia_intersect_gene_annotations <- mia_intersect_gene_annotations[!duplicated(mia_intersect_gene_annotations),]
#mia_match_intersect
mia_match_intersect_gene_annotations<- define_gene(insertion = mia_match_intersect_gene,
                                               genes = gene_annotation) 
mia_match_intersect_gene_annotations <- mia_match_intersect_gene_annotations[!duplicated(mia_match_intersect_gene_annotations),]
#mia_mismatch_intersect
mia_mismatch_intersect_gene_annotations<- define_gene(insertion = mia_mismatch_intersect_gene,
                                               genes = gene_annotation) 
mia_mismatch_intersect_gene_annotations <- mia_mismatch_intersect_gene_annotations[!duplicated(mia_mismatch_intersect_gene_annotations),]


##### SAVE  ##### 

# From mm1s match and mismatch cell line is removed. mm1s_intersect - is intersect of match and mismatch files without cell line
write.table(mm1s_intersect_gene_annotations, "../result/insertion_repeatmasker_annotation/mm1s_match_mismatch_interserct_no_cell_line_insertions_gene_annotations.txt", sep = "\t", quote = F, row.names = F)
# From mm1s match cell line and mismatch are removed
write.table(mm1s_match_intersect_gene_annotaions, "../result/insertion_repeatmasker_annotation/mm1s_match_unique_insertions_no_cell_line_mismatch_gene_annotaions.txt", sep = "\t", quote = F, row.names = F)
# From mm1s mismatch cell line and match are removed
write.table(mm1s_mismatch_intersect_gene_annotations, "../result/insertion_repeatmasker_annotation/mm1s_mismatch_unique_insertions_no_cell_line_match_gene_annotations.txt", sep = "\t", quote = F, row.names = F)

# From mia match and mismatch cell line is removed. mm1s_intersect - is intersect of match and mismatch files without cell line
write.table(mia_intersect_gene_annotations, "../result/insertion_repeatmasker_annotation/mia_match_mismatch_interserct_no_cell_line_insertions_gene_annotations.txt", sep = "\t", quote = F, row.names = F)
# From mia match cell line and mismatch are removed
write.table(mia_match_intersect_gene_annotations, "../result/insertion_repeatmasker_annotation/mia_match_unique_insertions_no_cell_line_mismatch_gene_annotations.txt", sep = "\t", quote = F, row.names = F)
# From mia mismatch cell line and match are removed
write.table(mia_mismatch_intersect_gene_annotations, "../result/insertion_repeatmasker_annotation/mia_mismatch_unique_insertions_no_cell_line_match__gene_annotations.txt", sep = "\t", quote = F, row.names = F)

```