---
title: "mutation_identification"
author: "Maria"
date: "1/10/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Functions
```{r functions}
check_nuc <- function(nuc)
{
  if(any(nuc == c("A", "T", "C", "G", "a", "t", "c", "g")))
  {
    return(T)
  }else
  {
    return(F)
  }
}
check_nuc_hetero <- function(nuc)
{
  if(any(nuc == c("M", "R", "W", "S", "Y", "K")))
  {
     return(T)
  }else
  {
    return(F)
  }
}


```

## MM1S mutations
Identification of the mutation

```{r mutations_mm1s}
#define paths and files
files_list <- dir(path = "../data/mm1s_results/alignment_with_repeats_ctdna_sep_without_contigs/modified/", full.names = T)


file_names <- dir(path = "../data/mm1s_results/alignment_with_repeats_ctdna_sep_without_contigs/modified/", full.names = F)

file_names <- sapply(file_names, function(x)
  {
  x <-paste( strsplit(x, ".fa")[[1]], ".txt", sep = "")
  return(x)
  })


for (i in seq_along(files_list)) 
  {
  aln <- readAAMultipleAlignment(files_list[i])
  print(file_names[i])
  mutations <- data.frame(ID = vector("character"),	
                          position  =vector("numeric"),	
                          nucleotide_in_contig = vector("character"),	
                          nucleotide_in_repeat =vector("character"))
  
  for (j in 1:unique(aln@unmasked@ranges@width)) 
    {
    if(length(aln@unmasked) == 8)
    {
          if(all(check_nuc(as.character(aln@unmasked[[2]][j])),
           check_nuc(as.character(aln@unmasked[[3]][j])),
           check_nuc(as.character(aln@unmasked[[4]][j])),
           check_nuc(as.character(aln@unmasked[[5]][j])),
           check_nuc(as.character(aln@unmasked[[6]][j])),
           check_nuc(as.character(aln@unmasked[[7]][j])),
           check_nuc(as.character(aln@unmasked[[8]][j]))))
    {
      contigs <- unique(toupper(c(as.character(aln@unmasked[[2]][j]),
                          as.character(aln@unmasked[[3]][j]),
                          as.character(aln@unmasked[[4]][j]),
                          as.character(aln@unmasked[[5]][j]),
                          as.character(aln@unmasked[[6]][j]),
                          as.character(aln@unmasked[[7]][j]))))
      if(length(contigs) == 1)
      {
        if(contigs != toupper(as.character(aln@unmasked[[8]][j])))
        {
          mutations <- rbind(mutations,
                             data.frame(ID = paste(strsplit(file_names[i], "_")[[1]][2],
                                               strsplit(file_names[i], "_")[[1]][3],
                                               strsplit(file_names[i], "_")[[1]][4],
                                               strsplit(file_names[i], "_")[[1]][5], sep = "_"),
                                        position = j,
                                        nucleotide_in_contig = contigs,
                                        nucleotide_in_repeat = as.character(aln@unmasked[[8]][j])))
          
        }
      }
    }
    }
    if(length(aln@unmasked) == 9)
          if(all(check_nuc(as.character(aln@unmasked[[3]][j])),
           check_nuc(as.character(aln@unmasked[[4]][j])),
           check_nuc(as.character(aln@unmasked[[5]][j])),
           check_nuc(as.character(aln@unmasked[[6]][j])),
           check_nuc(as.character(aln@unmasked[[7]][j])),
           check_nuc(as.character(aln@unmasked[[8]][j])),
           check_nuc(as.character(aln@unmasked[[9]][j]))))
    {
      contigs <- unique(toupper(c(as.character(aln@unmasked[[3]][j]),
                          as.character(aln@unmasked[[4]][j]),
                          as.character(aln@unmasked[[5]][j]),
                          as.character(aln@unmasked[[6]][j]),
                          as.character(aln@unmasked[[7]][j]),
                          as.character(aln@unmasked[[8]][j]))))
      if(length(contigs) == 1)
      {
        if(contigs != toupper(as.character(aln@unmasked[[9]][j])))
        {
          mutations <- rbind(mutations,
                             data.frame(ID = paste(strsplit(file_names[i], "_")[[1]][2],
                                               strsplit(file_names[i], "_")[[1]][3],
                                               strsplit(file_names[i], "_")[[1]][4],
                                               strsplit(file_names[i], "_")[[1]][5], sep = "_"),
                                        position = j,
                                        nucleotide_in_contig = contigs,
                                        nucleotide_in_repeat = as.character(aln@unmasked[[9]][j])))
          
        }
      }
    }
    }

  write.table(mutations,
              paste("../data/mm1s_results/alignment_with_repeats_ctdna_sep_without_contigs/mutation_index/",
                    strsplit(file_names[i], ".txt")[[1]],
                    "_mutation.txt", sep = ""), sep = "\t", row.names = F, quote = F)
  rm(mutations)
  
}

```


## MIA mutations

```{r mutations_mia}
#define paths and files
files_list <- dir(path = "../data/mia_results/mia_alignment_with_repeats_ctdna_sep_without_contigs/modified/", full.names = T)


file_names <- dir(path = "../data/mia_results/mia_alignment_with_repeats_ctdna_sep_without_contigs/modified/", full.names = F)


file_names <- sapply(file_names, function(x)
  {
  x <-paste( strsplit(x, ".fa")[[1]], ".txt", sep = "")
  return(x)
  })


for (i in seq_along(files_list)) 
  {
  aln <- readAAMultipleAlignment(files_list[i])
  print(file_names[i])
  mutations <- data.frame(ID = vector("character"),	
                          position  =vector("numeric"),	
                          nucleotide_in_contig = vector("character"),	
                          nucleotide_in_repeat =vector("character"))
  
  for (j in 1:unique(aln@unmasked@ranges@width)) 
    {
    if(length(aln@unmasked) == 13)
    {
          if(all(check_nuc(as.character(aln@unmasked[[2]][j])),
           check_nuc(as.character(aln@unmasked[[3]][j])),
           check_nuc(as.character(aln@unmasked[[4]][j])),
           check_nuc(as.character(aln@unmasked[[5]][j])),
           check_nuc(as.character(aln@unmasked[[6]][j])),
           check_nuc(as.character(aln@unmasked[[7]][j])),
           check_nuc(as.character(aln@unmasked[[8]][j])),
           check_nuc(as.character(aln@unmasked[[9]][j])),
           check_nuc(as.character(aln@unmasked[[10]][j])),
           check_nuc(as.character(aln@unmasked[[11]][j])),
           check_nuc(as.character(aln@unmasked[[12]][j])),
           check_nuc(as.character(aln@unmasked[[13]][j]))))
    {
      contigs <- unique(toupper(c(as.character(aln@unmasked[[2]][j]),
                          as.character(aln@unmasked[[3]][j]),
                          as.character(aln@unmasked[[4]][j]),
                          as.character(aln@unmasked[[5]][j]),
                          as.character(aln@unmasked[[6]][j]),
                          as.character(aln@unmasked[[8]][j]),
                          as.character(aln@unmasked[[9]][j]),
                          as.character(aln@unmasked[[10]][j]),
                          as.character(aln@unmasked[[11]][j]),
                          as.character(aln@unmasked[[12]][j]))))
      if(length(contigs) == 1)
      {
        if(contigs != toupper(as.character(aln@unmasked[[13]][j])))
        {
          mutations <- rbind(mutations,
                             data.frame(ID = paste(strsplit(file_names[i], "_")[[1]][2],
                                               strsplit(file_names[i], "_")[[1]][3],
                                               strsplit(file_names[i], "_")[[1]][4],
                                               strsplit(file_names[i], "_")[[1]][5], sep = "_"),
                                        position = j,
                                        nucleotide_in_contig = contigs,
                                        nucleotide_in_repeat = as.character(aln@unmasked[[13]][j])))
          
        }
      }
    }
    }
    if(length(aln@unmasked) == 14)
    {
          if(all(check_nuc(as.character(aln@unmasked[[3]][j])),
           check_nuc(as.character(aln@unmasked[[4]][j])),
           check_nuc(as.character(aln@unmasked[[5]][j])),
           check_nuc(as.character(aln@unmasked[[6]][j])),
           check_nuc(as.character(aln@unmasked[[7]][j])),
           check_nuc(as.character(aln@unmasked[[8]][j])),
           check_nuc(as.character(aln@unmasked[[9]][j])),
           check_nuc(as.character(aln@unmasked[[10]][j])),
           check_nuc(as.character(aln@unmasked[[11]][j])),
           check_nuc(as.character(aln@unmasked[[12]][j])),
           check_nuc(as.character(aln@unmasked[[13]][j])),
           check_nuc(as.character(aln@unmasked[[14]][j]))))
    {
      contigs <- unique(toupper(c(as.character(aln@unmasked[[3]][j]),
                          as.character(aln@unmasked[[4]][j]),
                          as.character(aln@unmasked[[5]][j]),
                          as.character(aln@unmasked[[6]][j]),
                          as.character(aln@unmasked[[8]][j]),
                          as.character(aln@unmasked[[9]][j]),
                          as.character(aln@unmasked[[10]][j]),
                          as.character(aln@unmasked[[11]][j]),
                          as.character(aln@unmasked[[12]][j]),
                          as.character(aln@unmasked[[13]][j]))))
      if(length(contigs) == 1)
      {
        if(contigs != toupper(as.character(aln@unmasked[[14]][j])))
        {
          mutations <- rbind(mutations,
                             data.frame(ID = paste(strsplit(file_names[i], "_")[[1]][2],
                                               strsplit(file_names[i], "_")[[1]][3],
                                               strsplit(file_names[i], "_")[[1]][4],
                                               strsplit(file_names[i], "_")[[1]][5], sep = "_"),
                                        position = j,
                                        nucleotide_in_contig = contigs,
                                        nucleotide_in_repeat = as.character(aln@unmasked[[14]][j])))
          
        }
      }
    }
    }
  }
  

  write.table(mutations,
              paste("../data/mia_results/mia_alignment_with_repeats_ctdna_sep_without_contigs/mutation_index/",
                    strsplit(file_names[i], ".txt")[[1]],
                    "_mutation.txt", sep = ""), sep = "\t", row.names = F, quote = F)
  rm(mutations)
  
}

```


##### Extract TE fasta files for MM1S
```{r extract_fasta_mm1s}
### without substutution of the heterozygous genoytpes
consensus <- readAAStringSet("../data/mm1s_results/blast_ctdn_consensus/mm1s_ctdna_repeats_consensus_full_final.fa")

files_list_alignment <- dir(path = "../data/mm1s_results/alignment_with_repeats_ctdna_sep_without_contigs/modified/", full.names = T)


files_list_mutation <- dir(path = "../data/mm1s_results/te_mutation_index", full.names = T)


files_list_mutation_names <-  dir(path = "../data/mm1s_results/te_mutation_index", full.names = F)
files_list_mutation_names <- sapply(files_list_mutation_names, function(x)
  {
  x <-strsplit(x, ".txt")[[1]][1]
  return(x)
  })

for (i in seq_along(files_list_mutation_names)) 
  {
  aln <- readAAMultipleAlignment(files_list_alignment[grep(files_list_mutation_names[i],files_list_alignment )])
  if(unique(aln@unmasked@ranges@width) == consensus@ranges@width[consensus@ranges@NAMES == files_list_mutation_names[i]])
  {
    if(length(aln@unmasked) == 8)
    {
      te_repeats <- aln@unmasked[[8]]
      consensus_tmp <- consensus[[which(consensus@ranges@NAMES == files_list_mutation_names[i])]]
      for (j in 1:length(te_repeats)) 
        {
        if(as.character(te_repeats[j] )!= "-")
        {
          te_repeats[j] <- consensus_tmp[j]
        }
      }
      # add orignial req
      te_repeats_original <-  aln@unmasked[[8]]
      te_repeats_original <- gsub("-", "", as.character(te_repeats_original))
      
      
      te_repeats_nogaps <- gsub("-", "", as.character(te_repeats))
      te_repeats_nogaps <- paste(">",files_list_mutation_names[i],"_mutated",  "\n", te_repeats_nogaps, "\n",
                                 ">", files_list_mutation_names[i],"_reference", "\n", te_repeats_original, "\n", sep = "")
      write(te_repeats_nogaps,
            paste("../data/mm1s_results/te_sequences/",files_list_mutation_names[i], ".fa", sep = ""), ncolumns = 1)
    }
    if(length(aln@unmasked) == 9)
    {
      te_repeats <- aln@unmasked[[9]]
      consensus_tmp <- consensus[[which(consensus@ranges@NAMES == files_list_mutation_names[i])]]
      for (j in 1:length(te_repeats)) 
        {
        if(as.character(te_repeats[j] )!= "-")
        {
          te_repeats[j] <- consensus_tmp[j]
        }
      }
      # add orignial req
      te_repeats_original <-  aln@unmasked[[9]]
      te_repeats_original <- gsub("-", "", as.character(te_repeats_original))
      
      
      te_repeats_nogaps <- gsub("-", "", as.character(te_repeats))
      te_repeats_nogaps <- paste(">",files_list_mutation_names[i], "_mutated", "\n", te_repeats_nogaps, "\n",
                                 ">", files_list_mutation_names[i],"_reference", "\n", te_repeats_original, "\n", sep = "")
      write(te_repeats_nogaps,
            paste("../data/mm1s_results/te_sequences/",files_list_mutation_names[i], ".fa", sep = ""), ncolumns = 1)
    }

  }else
  {
     print(paste("Failed ", files_list_mutation_names[i]))
  }
   
}


### with substutution of the heterozygous genoytpes
consensus <- readAAStringSet("../data/mm1s_results/blast_ctdn_consensus/mm1s_ctdna_repeats_consensus_full_final.fa")

files_list_alignment <- dir(path = "../data/mm1s_results/alignment_with_repeats_ctdna_sep_without_contigs/modified/", full.names = T)
files_list_alignment <- files_list_alignment[-grep("desktop", files_list_alignment)]

files_list_mutation <- dir(path = "../data/mm1s_results/mm1s_final_report/te_mutation_index", full.names = T)
files_list_mutation <- files_list_alignment[-grep("desktop", files_list_mutation)]

files_list_mutation_names <-  dir(path = "../data/mm1s_results/mm1s_final_report/te_mutation_index", full.names = F)
files_list_mutation_names <- files_list_mutation_names[-grep("desktop", files_list_mutation_names)]
files_list_mutation_names <- sapply(files_list_mutation_names, function(x)
  {
  x <-strsplit(x, ".txt")[[1]][1]
  return(x)
  })

### define heterozygous codes
nucs <- list(M = c("A", "C"),
             R = c("A", "G"),
             W = c("A", "T"),
             S = c("C", "G"),
             Y = c("C", "T"),
             K = c("G", "T"))


for (i in seq_along(files_list_mutation_names)) 
  {
  print(files_list_mutation_names[i])
  aln <- readAAMultipleAlignment(files_list_alignment[grep(files_list_mutation_names[i],files_list_alignment )])
  if(unique(aln@unmasked@ranges@width) == consensus@ranges@width[consensus@ranges@NAMES == files_list_mutation_names[i]])
  {
    if(length(aln@unmasked) == 8)
    {

      te_repeats <- aln@unmasked[[8]]
      te_repeats_original <-  aln@unmasked[[8]]
    
      consensus_tmp <- consensus[[which(consensus@ranges@NAMES == files_list_mutation_names[i])]]
      
      for (j in 1:length(te_repeats)) 
        {
        te_repeats_nuc <- toupper(as.character(te_repeats[j]))
        consensus_tmp_nuc <- toupper(as.character(consensus_tmp[j]))
        ctdna <-toupper(c(as.character(aln@unmasked[[2]][j]),
                       as.character(aln@unmasked[[3]][j]),
                       as.character(aln@unmasked[[4]][j]),
                       as.character(aln@unmasked[[5]][j]),
                       as.character(aln@unmasked[[6]][j]),
                      as.character(aln@unmasked[[7]][j])))

        if(te_repeats_nuc != "-" & check_nuc(consensus_tmp_nuc))
        {
          te_repeats[j] <- consensus_tmp_nuc
        }
        if(te_repeats_nuc != "-" & check_nuc_hetero(consensus_tmp_nuc))
        {
          if(te_repeats_nuc != ctdna[6] & check_nuc(ctdna[6]))
          {
            if(length(unique(ctdna[1:5])) <= 2)
            {
               te_repeats[j] <- ctdna[6]
            }else
            {
               te_repeats[j] <- te_repeats_nuc
            }
              
          }
        }
      }
    }
    
    if(length(aln@unmasked) == 9)
    {

      te_repeats <- aln@unmasked[[9]]
      te_repeats_original <-  aln@unmasked[[9]]
    
      consensus_tmp <- consensus[[which(consensus@ranges@NAMES == files_list_mutation_names[i])]]
      
      for (j in 1:length(te_repeats)) 
        {
        te_repeats_nuc <- toupper(as.character(te_repeats[j]))
        consensus_tmp_nuc <- toupper(as.character(consensus_tmp[j]))
        ctdna <-toupper(c(as.character(aln@unmasked[[3]][j]),
                       as.character(aln@unmasked[[4]][j]),
                       as.character(aln@unmasked[[5]][j]),
                       as.character(aln@unmasked[[6]][j]),
                      as.character(aln@unmasked[[7]][j]),
                      as.character(aln@unmasked[[8]][j])))

        if(te_repeats_nuc != "-" & check_nuc(consensus_tmp_nuc))
        {
          te_repeats[j] <- consensus_tmp_nuc
        }
        if(te_repeats_nuc != "-" & check_nuc_hetero(consensus_tmp_nuc))
        {
          if(te_repeats_nuc != ctdna[6] & check_nuc(ctdna[6]))
          {
            if(length(unique(ctdna[1:5])) <= 2)
            {
               te_repeats[j] <- ctdna[6]
            }else
            {
               te_repeats[j] <- te_repeats_nuc
            }
              
          }
        }
      }
    }
    
      # Remove gaps
      te_repeats_original <- gsub("-", "", as.character(te_repeats_original))
      
      te_repeats_nogaps <- gsub("-", "", as.character(te_repeats))
      te_repeats_nogaps <- paste(">",files_list_mutation_names[i],"_mutated",  "\n", te_repeats_nogaps, "\n",
                                 ">", files_list_mutation_names[i],"_reference", "\n", te_repeats_original, "\n", sep = "")
      write(te_repeats_nogaps,
            paste("../data/mm1s_results/mm1s_final_report/te_sequences_without_hetero/",files_list_mutation_names[i], ".fa", sep = ""), ncolumns = 1)
    }
 
   
}
```

##### Extract TE fasta files for MIA
```{r extract_fasta_mia}
consensus <- readAAStringSet("../data/mia_results/blast_ctdn_consensus/mia_ctdna_repeat_consensus.txt")

files_list_alignment <- dir(path = "../data/mia_results/mia_alignment_with_repeats_ctdna_sep_without_contigs/modified/", full.names = T)


files_list_mutation <- dir(path = "../data/mia_results/mia_alignment_with_repeats_ctdna_sep_without_contigs/mutation_index/", full.names = T)


files_list_mutation_names <-  dir(path = "../data/mia_results/mia_alignment_with_repeats_ctdna_sep_without_contigs/mutation_index/", full.names = F)

files_list_mutation_names <- sapply(files_list_mutation_names, function(x)
  {
  x <-paste(strsplit(x, "_")[[1]][2:5], collapse = "_")
  return(x)
  })

for (i in seq_along(files_list_mutation_names)) 
  {
  aln <- readAAMultipleAlignment(files_list_alignment[i])
  if(unique(aln@unmasked@ranges@width) == consensus@ranges@width[consensus@ranges@NAMES == files_list_mutation_names[i]])
  {
    if(length(aln@unmasked) == 13)
    {
      te_repeats <- aln@unmasked[[13]]
      consensus_tmp <- consensus[[which(consensus@ranges@NAMES == files_list_mutation_names[i])]]
      for (j in 1:length(te_repeats))
        {
        if(as.character(te_repeats[j] )!= "-")
        {
          te_repeats[j] <- consensus_tmp[j]
        }
      }
      # add orignial req
      te_repeats_original <-  aln@unmasked[[13]]
      te_repeats_original <- gsub("-", "", as.character(te_repeats_original))


      te_repeats_nogaps <- gsub("-", "", as.character(te_repeats))
      te_repeats_nogaps <- paste(">",files_list_mutation_names[i],"_mutated",  "\n", te_repeats_nogaps, "\n",
                                 ">", files_list_mutation_names[i],"_reference", "\n", te_repeats_original, "\n", sep = "")
      write(te_repeats_nogaps,
            paste("../data/mia_results/te_sequences_mia/",files_list_mutation_names[i], ".fa", sep = ""), ncolumns = 1)
    }
    if(length(aln@unmasked) == 14)
    {
      te_repeats <- aln@unmasked[[14]]
      consensus_tmp <- consensus[[which(consensus@ranges@NAMES == files_list_mutation_names[i])]]
      for (j in 1:length(te_repeats))
        {
        if(as.character(te_repeats[j] )!= "-")
        {
          te_repeats[j] <- consensus_tmp[j]
        }
      }
      # add orignial req
      te_repeats_original <-  aln@unmasked[[14]]
      te_repeats_original <- gsub("-", "", as.character(te_repeats_original))


      te_repeats_nogaps <- gsub("-", "", as.character(te_repeats))
      te_repeats_nogaps <- paste(">",files_list_mutation_names[i], "_mutated", "\n", te_repeats_nogaps, "\n",
                                 ">", files_list_mutation_names[i],"_reference", "\n", te_repeats_original, "\n", sep = "")
      write(te_repeats_nogaps,
            paste("../data/mia_results/te_sequences_mia/",files_list_mutation_names[i], ".fa", sep = ""), ncolumns = 1)
    }

  }else
  {
     print(paste("Failed ", i,files_list_mutation_names[i]))
  }
   
}


### with substutution of the heterozygous genoytpes
consensus <- readAAStringSet("../data/mia_results/blast_ctdn_consensus/mia_ctdna_repeat_consensus.fa")

files_list_alignment <- dir(path = "../data/mia_results/mia_alignment_with_repeats_ctdna_sep_without_contigs/modified/", full.names = T)
files_list_alignment <- files_list_alignment[-grep("desktop", files_list_alignment)]

files_list_mutation <- dir(path = "../data/mia_results/mia_alignment_with_repeats_ctdna_sep_without_contigs/mutation_index/", full.names = T)
files_list_mutation <- files_list_mutation[-grep("desktop", files_list_mutation)]

files_list_mutation_names <-  dir(path = "../data/mia_results/mia_alignment_with_repeats_ctdna_sep_without_contigs/mutation_index/", full.names = F)
files_list_mutation_names <- files_list_mutation_names[-grep("desktop", files_list_mutation_names)]
files_list_mutation_names <- sapply(files_list_mutation_names, function(x)
  {
  x <-paste(strsplit(x, "_")[[1]][2:5], collapse = "_")
  return(x)
  })
### define heterozygous codes
nucs <- list(M = c("A", "C"),
             R = c("A", "G"),
             W = c("A", "T"),
             S = c("C", "G"),
             Y = c("C", "T"),
             K = c("G", "T"))


for (i in seq_along(files_list_mutation_names)) 
  {
  print(files_list_mutation_names[i])
  aln <- readAAMultipleAlignment(files_list_alignment[grep(files_list_mutation_names[i],files_list_alignment )])
  if(unique(aln@unmasked@ranges@width) == consensus@ranges@width[consensus@ranges@NAMES == files_list_mutation_names[i]])
  {
    if(length(aln@unmasked) == 13)
    {

      te_repeats <- aln@unmasked[[13]]
      te_repeats_original <-  aln@unmasked[[13]]
    
      consensus_tmp <- consensus[[which(consensus@ranges@NAMES == files_list_mutation_names[i])]]
      
      for (j in 1:length(te_repeats)) 
        {
        te_repeats_nuc <- toupper(as.character(te_repeats[j]))
        consensus_tmp_nuc <- toupper(as.character(consensus_tmp[j]))
        ctdna <-toupper(c(as.character(aln@unmasked[[2]][j]),
                       as.character(aln@unmasked[[3]][j]),
                       as.character(aln@unmasked[[4]][j]),
                       as.character(aln@unmasked[[5]][j]),
                       as.character(aln@unmasked[[6]][j]),
                      as.character(aln@unmasked[[7]][j]),
                      as.character(aln@unmasked[[8]][j]),
                      as.character(aln@unmasked[[9]][j]),
                      as.character(aln@unmasked[[10]][j]),
                      as.character(aln@unmasked[[11]][j]),
                      as.character(aln@unmasked[[12]][j])))

        if(te_repeats_nuc != "-" & check_nuc(consensus_tmp_nuc))
        {
          te_repeats[j] <- consensus_tmp_nuc
        }
        if(te_repeats_nuc != "-" & check_nuc_hetero(consensus_tmp_nuc))
        {
          if(te_repeats_nuc != ctdna[11] & check_nuc(ctdna[11]) )
          {
            if(length(unique(ctdna[1:10])) <= 2)
            {
               te_repeats[j] <- ctdna[11]
            }else
            {
               te_repeats[j] <- te_repeats_nuc
            }
              
          }
        }
      }
    }
    
    if(length(aln@unmasked) == 14)
    {

      te_repeats <- aln@unmasked[[14]]
      te_repeats_original <-  aln@unmasked[[14]]
    
      consensus_tmp <- consensus[[which(consensus@ranges@NAMES == files_list_mutation_names[i])]]
      
      for (j in 1:length(te_repeats)) 
        {
        te_repeats_nuc <- toupper(as.character(te_repeats[j]))
        consensus_tmp_nuc <- toupper(as.character(consensus_tmp[j]))
        ctdna <-toupper(c(as.character(aln@unmasked[[3]][j]),
                       as.character(aln@unmasked[[4]][j]),
                       as.character(aln@unmasked[[5]][j]),
                       as.character(aln@unmasked[[6]][j]),
                      as.character(aln@unmasked[[7]][j]),
                      as.character(aln@unmasked[[8]][j]),
                      as.character(aln@unmasked[[9]][j]),
                      as.character(aln@unmasked[[10]][j]),
                      as.character(aln@unmasked[[11]][j]),
                      as.character(aln@unmasked[[12]][j]),
                      as.character(aln@unmasked[[13]][j])))

        if(te_repeats_nuc != "-" & check_nuc(consensus_tmp_nuc))
        {
          te_repeats[j] <- consensus_tmp_nuc
        }
        if(te_repeats_nuc != "-" & check_nuc_hetero(consensus_tmp_nuc))
        {
          if(te_repeats_nuc != ctdna[11] & check_nuc(ctdna[11]))
          {
            if(length(unique(ctdna[1:10])) <= 2)
            {
               te_repeats[j] <- ctdna[11]
            }else
            {
               te_repeats[j] <- te_repeats_nuc
            }
              
          }
        }
      }
    }
    
      # Remove gaps
      te_repeats_original <- gsub("-", "", as.character(te_repeats_original))
      
      te_repeats_nogaps <- gsub("-", "", as.character(te_repeats))
      te_repeats_nogaps <- paste(">",files_list_mutation_names[i],"_mutated",  "\n", te_repeats_nogaps, "\n",
                                 ">", files_list_mutation_names[i],"_reference", "\n", te_repeats_original, "\n", sep = "")
      write(te_repeats_nogaps,
            paste("../data/mia_results/te_sequences_without_hetero_mia/",files_list_mutation_names[i], ".fa", sep = ""), ncolumns = 1)
    }
 
   
}

```

#### Transposone mutation indentification MM1S
define the mutation position in the transposon

```{r trasposone_mutations}

files_list <- dir(path = "../data/mm1s_results/mm1s_final_report/te_sequences_without_hetero", full.names = T)



files_list_names <- dir(path = "../data/mm1s_results/mm1s_final_report/te_sequences_without_hetero", full.names = F)


files_list_names <- sapply(files_list_names, function(x)
  {
  x <-strsplit(x, ".fa")[[1]][1]
  return(x)
  })
transposone_location <- readxl::read_excel("../data/mm1s_results/mm1s_final_report/mm1s_transposon_location.xlsx", sheet = 1)


for (i in seq_along(files_list)) 
  {
  print(i)
  mutation_index <- data.frame(ransposone_ID = vector("character"),
                             mutaion_position = vector("numeric"),
                             nucleotide_in_contig =  vector("character"),
                             nucleotide_in_reference =  vector("character"))
  
  te_seq <- readDNAStringSet(files_list[i])
  te_transposone_location <- transposone_location$`Transposone orientation`[which(transposone_location$Transposone_ID == files_list_names[i])]
  
  if(length(te_transposone_location) > 0)
  {
      if(te_transposone_location == "reverse complement")
  {
   te_seq <- reverseComplement(te_seq) 
  }
  
     for (j in 1:unique(te_seq@ranges@width)) 
     {
     if(as.character(te_seq[[1]][j]) != as.character(te_seq[[2]][j]))
     {
      mutation_index_temp <- data.frame(ransposone_ID = files_list_names[i],
                             mutaion_position = j,
                             nucleotide_in_contig =  as.character(te_seq[[1]][j]),
                             nucleotide_in_reference = as.character(te_seq[[2]][j]))
      mutation_index <- rbind(mutation_index, mutation_index_temp)
     }

     }
 
  mutation_index_temp$nucleotide_in_contig <- as.character(mutation_index_temp$nucleotide_in_contig)
  mutation_index_temp$nucleotide_in_reference <- as.character(mutation_index_temp$nucleotide_in_reference)
  writeXStringSet(te_seq, paste("../data/mm1s_results/mm1s_final_report/te_sequences_reversed/", files_list_names[i], ".fa", sep = ""))
  write.table(mutation_index,
              paste("../data/mm1s_results/mm1s_final_report/te_reversed_mutation_index/",  files_list_names[i], "_mutation_locatoin_direct.txt", sep = ""), sep = "\t", quote = F, row.names = F)
  }
  

}

```

#### Transposone mutation indentification MIA
define the mutation position in the transposon

```{r trasposone_mutations_mia}

files_list <- dir(path = "../data/mia_results/mia_final_report/te_sequences_without_hetero_mia", full.names = T)

files_list_names <- dir(path = "../data/mia_results/mia_final_report/te_sequences_without_hetero_mia", full.names = F)

files_list_names <- sapply(files_list_names, function(x)
  {
  x <-strsplit(x, ".fa")[[1]][1]
  return(x)
  })
transposone_location <- readxl::read_excel("../data/mia_results/mia_final_report/mia_transposon_location.xlsx", sheet = 1)


for (i in seq_along(files_list)) 
  {
  print(i)
  mutation_index <- data.frame(ransposone_ID = vector("character"),
                             mutaion_position = vector("numeric"),
                             nucleotide_in_contig =  vector("character"),
                             nucleotide_in_reference =  vector("character"))
  
  te_seq <- readDNAStringSet(files_list[i])
  te_transposone_location <- transposone_location$`Transposone orientation`[which(transposone_location$Transposone_ID == files_list_names[i])]
  
  if(length(te_transposone_location) > 0)
  {
      if(te_transposone_location == "reverse complement")
  {
   te_seq <- reverseComplement(te_seq) 
  }
  
     for (j in 1:unique(te_seq@ranges@width)) 
     {
     if(as.character(te_seq[[1]][j]) != as.character(te_seq[[2]][j]))
     {
      mutation_index_temp <- data.frame(ransposone_ID = files_list_names[i],
                             mutaion_position = j,
                             nucleotide_in_contig =  as.character(te_seq[[1]][j]),
                             nucleotide_in_reference = as.character(te_seq[[2]][j]))
      mutation_index <- rbind(mutation_index, mutation_index_temp)
     }

     }
 
  mutation_index_temp$nucleotide_in_contig <- as.character(mutation_index_temp$nucleotide_in_contig)
  mutation_index_temp$nucleotide_in_reference <- as.character(mutation_index_temp$nucleotide_in_reference)
  writeXStringSet(te_seq, paste("../data/mia_results/mia_final_report/te_sequences_reversed/", files_list_names[i], ".fa", sep = ""))
  write.table(mutation_index,
              paste("../data/mia_results/mia_final_report/te_reversed_mutation_index/",  files_list_names[i], "_mutation_locatoin_direct.txt", sep = ""), sep = "\t", quote = F, row.names = F)
  }
  

}

```


